requirements/prod.txt

```diff
@@ -1,6 +1,7 @@
 boto3>=1.3.0,<2
 click>=7.0,<9.0
 colorama>=0.2.5,<0.4.4
+dictdiffer==0.8.1
```

Comment:
have you looked at DeepDiff? In my experience it's a lot more robust of a diffing library. It can be very exact about what has changed and how it changed. It even has useful serialization formats that are fairly easy to read.

Reply:
So in this implementation the default diffing library we use is difflib.unified_diff. We found that gives us the most easily digestible format for humans. There are some issues with it though, specifically around how it handles ordering with JSON templates. Also I'm pretty sure it's only python 3 (but is that really a problem? 😛 ).

Note that a feature based on difflib.unified_diff was added upstream in #212 but then reverted again later in #225. 

For those 2 reasons we also wanted to provide an option for a much more robust option if the user required it.

We did look into using DeepDiff, however, found that dictdiffer provided extra information in a format that could more easily be parsed. YMMV.  If DeepDiff can be configured to give us a friendly diff format similar to difflib.unified_diff, that would be the best of both worlds!

Examples Below:

difflib.unified_diff.
```diff
test-test: --- remote_template
+++ local_template
@@ -23,10 +23,14 @@
         - Key: Project
           Value: GitOps
+        - Key: Foo
+          Value: Bar
 
 Outputs:
   InstanceArn:
     Value: !Ref MyInstance
+  AvailabilityZone:
+    Value: !GetAtt MyInstance.AvailabilityZone
 
 # vim: set ft=yaml:
```

dictdiffer
```
[aws:personal1] /src $ sceptre --dir=test --var-file=/env/common-env.yaml --var-file=/env/test.yaml diff test.yaml 
["test-test: [('add', 'Resources.MyInstance.Properties.Tags', [(6, {'Key': 'Foo', 'Value': 'Bar'})]), ('remove', 'Resources.MyInstance.Properties', [('InstanceType', 'm1.small')])]"]
```

DeepDiff
```
[aws:personal1] /src $ sceptre --dir=test --var-file=/env/common-env.yaml --var-file=/env/test.yaml diff test.yaml 
test-test: Item root['Resources']['MyInstance']['Properties']['InstanceType'] removed from dictionary.
Item root['Resources']['MyInstance']['Properties']['Tags'][6] added to iterable.
```

Reply:
I found it a lot easier to read in the template file into a dictionary and do the diff on the dictionary itself rather than doing the diff on the templates. AWS's cfn-flip library lets you load a template, regardless of whether it is in json or yaml and whether it has `!Ref` or things like it in it. It loads it as a dictionary that can be compared key-for-key. This was really useful and tackles the resource and parameter ordering issue.

Take a look at #1093 and the output I posted there. I'm essentially doing this process:
1. Load the template to a dict with cfn_flip.load
2. diff the two with DeepDiff(deployed, generated, verbosity=2) (the verbosity of 2 shows exactly what the different values are so it's not just "item added" and "item removed"
3. I serialize the value to a readable json using `diff.to_json()` (which has some type serializatin functionality). If the sceptre output_format isn't "json", I convert the json to yaml.

This produces something like this:
```yaml
type_changes:
  root['Resources']['BucketPolicy']['Properties']['PolicyDocument']['Statement'][0]['Action']:
    new_type: list
    new_value:
      - s3:GetObject
      - s3:PutObject
    old_type: str
    old_value: s3:GetObject
```